gix-blame/src/file/function.rs

```diff
@@ -322,15 +327,121 @@ fn tree_diff_at_file_path(
     let tree_iter = odb.find_tree_iter(&tree_id, rhs_tree_buf)?;
     stats.trees_decoded += 1;
 
-    let mut recorder = gix_diff::tree::Recorder::default();
-    gix_diff::tree(parent_tree_iter, tree_iter, state, &odb, &mut recorder)?;
+    let mut recorder = Recorder::new(file_path.into());
+    // TODO
+    // `recorder` cancels the traversal by returning `Cancel` when a change to `file_path` is
+    // found. `gix_diff::tree` converts `Cancel` into `Err(...)` which is why we ignore its return
+    // value here. I don’t know whether this has the potential to hide bugs.
+    let _ = gix_diff::tree(parent_tree_iter, tree_iter, state, &odb, &mut recorder);
     stats.trees_diffed += 1;
 
-    Ok(recorder.records.into_iter().find(|change| match change {
-        gix_diff::tree::recorder::Change::Modification { path, .. } => path == file_path,
-        gix_diff::tree::recorder::Change::Addition { path, .. } => path == file_path,
-        gix_diff::tree::recorder::Change::Deletion { path, .. } => path == file_path,
-    }))
+    Ok(recorder.change)
+}
+
+// TODO
+// The name is preliminary and can potentially include more context. Also, this should probably be
+// moved to its own location.
+struct Recorder {
+    interesting_path: BString,
+    path_deque: VecDeque<BString>,
+    path: BString,
+    change: Option<gix_diff::tree::recorder::Change>,
+}
+
+impl Recorder {
+    fn new(interesting_path: BString) -> Self {
+        Recorder {
+            interesting_path,
+            path_deque: Default::default(),
+            path: Default::default(),
+            change: None,
+        }
+    }
+
+    fn pop_element(&mut self) {
```

Comment:
These utilities can be removed if one would have an `inner: Recorder` which does the bookkeeping. Then one can focus on `visit` and implement what matters.

Reply:
I’ve made `Recorder` a wrapper around `inner: gix_diff::tree::Recorder` (I hope I understood your suggestion correctly). This might have come at a minimal cost to performance, but that impression might also have been noise in running `hyperfine`.